### ISSUE TITLE
[C++] Add MemoryPool API to resize without copying the data

### ISSUE THREAD

**Issue by pitrou** [apache/arrow#40686]
### Describe the enhancement requested

A common memory (re)allocation pattern is to persist a dynamically-sized scratch buffer (for example when decompressing or decoding data). Currently, the two options to achieve that are suboptimal:
1. either resize the buffer when needed, but if the buffer's base pointer changes, the data will be copied when resizing even though the data is ephemeral
2. or allocate a new buffer when needing to increase the scratch size, but it is wasteful if the buffer could otherwise have been resized in place

Both jemalloc and mimalloc expose APIs that would allow us to implement a more optimal behavior of resizing without copying existing data: [mi_expand](https://microsoft.github.io/mimalloc/group__malloc.html#gaaee66a1d483c3e28f585525fb96707e4) for mimalloc, [xallocx](https://jemalloc.net/jemalloc.3.html) for jemalloc.

One possible API would be:
```c++
  Status ReallocateNoCopy(int64_t old_size, int64_t new_size, int64_t alignment,
                          uint8_t** ptr) = 0;
  Status ReallocateNoCopy(int64_t old_size, int64_t new_size, uint8_t** ptr);
```

### Component(s)

C++

**Comment by pitrou** [comment]
@mapleFU @felipecrv

**Comment by mapleFU** [comment]
Will glance at malloc(3)'s api.

>  allocate a new buffer when needing to increase the scratch size, but it is wasteful if the buffer could otherwise have been resized in place

Whats the difference between this and not call shrink_to_fit when goes to smaller?

**Comment by felipecrv** [comment]
These functions are very likely to fail.

> The xallocx() function returns the real size of the resulting resized allocation pointed to by ptr, which is a value less than size if the allocation could not be adequately grown in place.

I don't understand why using these to implement `MemoryPool::Reallocate` is not enough.

**Comment by pitrou** [comment]
> These functions are very likely to fail.

Yes!

> I don't understand why using these to implement `MemoryPool::Reallocate` is not enough.

Because I'm proposing another API :-)

**Comment by felipecrv** [comment]
> Because I'm proposing another API :-)

Right, but let's say it's being used by a `Builder`, should the resize errors bubble up to the users of the builder?

`Resize()` on builders can fail (but very unlikely) and are implemented in terms of `Reallocate`. Would this new API affect how `Builder::Resize()` is implemented or your would add `Builder::TryResizeInPlace()`?

**Comment by pitrou** [comment]
Again, no, what I'm proposing to expose is a `ReallocateNoCopy` primitive.
